File: android/src/main/java/com/lebrislo/bluetooth/mesh/plugin/PluginCallManager.kt
package com.lebrislo.bluetooth.mesh.plugin

import android.util.Log
import com.getcapacitor.PluginCall
import com.lebrislo.bluetooth.mesh.BluetoothMeshPlugin.Companion.MESH_MODEL_MESSAGE_EVENT_STRING
import com.lebrislo.bluetooth.mesh.plugin.FoundationPluginCall.Companion.generateFoundationPluginCallResponse
import com.lebrislo.bluetooth.mesh.plugin.OperationPair.Companion.getOperationPair
import com.lebrislo.bluetooth.mesh.plugin.SigPluginCall.Companion.generateSigPluginCallResponse
import com.lebrislo.bluetooth.mesh.plugin.VendorPluginCall.Companion.generateVendorPluginCallResponse
import com.lebrislo.bluetooth.mesh.utils.NotificationManager
import no.nordicsemi.android.mesh.transport.MeshMessage
import no.nordicsemi.android.mesh.utils.MeshAddress

/**
 * This class is used to manage plugin calls.
 */
class PluginCallManager private constructor() {
    private val tag: String = PluginCallManager::class.java.simpleName

    private val pluginCalls: MutableList<BasePluginCall> = mutableListOf()

    companion object {

        @Volatile
        private var instance: PluginCallManager? = null

        fun getInstance() =
            instance ?: synchronized(this) {
                instance ?: PluginCallManager().also { instance = it }
            }
    }

    /**
     * Remove a plugin call from the list of plugin calls.
     *
     * @param call Plugin call.
     */
    fun removePluginCall(call: BasePluginCall) {
        pluginCalls.remove(call)
        Log.d(tag, "Remove call due to timeout ${call.call.methodName}")
    }

    /**
     * Add a SIG plugin call to the list of plugin calls to watch for a response.
     *
     * @param meshOperation Mesh operation.
     * @param meshAddress Mesh address.
     * @param call Plugin call.
     */
    fun addSigPluginCall(meshOperation: Int, meshAddress: Int, call: PluginCall) {
        if (!MeshAddress.isValidUnicastAddress(meshAddress)) {
            return call.resolve()
        }
        val operationPair = getOperationPair(meshOperation)
        pluginCalls.add(SigPluginCall(operationPair, meshAddress, call))
    }

    /**
     * Resolve a SIG plugin call.
     * If the call is not found in the list of plugin calls, a notification is sent to the listeners.
     *
     * @param meshMessage Mesh message.
     */
    fun resolveSigPluginCall(meshMessage: MeshMessage) {
        Log.d(tag, "resolveSigPluginCall ${meshMessage.opCode} from ${meshMessage.src}")
        val callResponse = generateSigPluginCallResponse(meshMessage)

        val pluginCall =
            pluginCalls.find { it is SigPluginCall && it.meshOperationCallback == meshMessage.opCode && it.meshAddress == meshMessage.src }

        Log.d(tag, "resolveSigPluginCall: registered call: ${pluginCall != null}")

        if (pluginCall != null) {
            pluginCall as SigPluginCall
            pluginCall.resolve(callResponse)
            pluginCalls.remove(pluginCall)
        }
        NotificationManager.getInstance().sendNotification(MESH_MODEL_MESSAGE_EVENT_STRING, callResponse)
    }

    /**
     * Add a Foundation plugin call to the list of plugin calls to watch for a response.
     *
     * @param meshOperation Mesh operation.
     * @param meshAddress Mesh address.
     * @param call Plugin call.
     */
    fun addFoundationPluginCall(meshOperation: Int, meshAddress: Int, call: PluginCall) {
        val operationPair = getOperationPair(meshOperation)
        pluginCalls.add(FoundationPluginCall(operationPair, meshAddress, call))
    }

    /**
     * Resolve a Foundation plugin call.
     * If the call is not found in the list of plugin calls, a notification is sent to the listeners.
     *
     * @param meshMessage Mesh message.
     */
    fun resolveFoundationPluginCall(meshMessage: MeshMessage) {
        Log.d(tag, "resolveFoundationPluginCall ${meshMessage.opCode} from ${meshMessage.src}")
        val callResponse = generateFoundationPluginCallResponse(meshMessage)

        val pluginCall =
            pluginCalls.find { it is FoundationPluginCall && it.meshOperationCallback == meshMessage.opCode && it.meshAddress == meshMessage.src }

        Log.d(tag, "resolveFoundationPluginCall: registered call: ${pluginCall != null}")

        if (pluginCall != null) {
            pluginCall as FoundationPluginCall
            pluginCall.resolve(callResponse)
            pluginCalls.remove(pluginCall)
        }
        NotificationManager.getInstance().sendNotification(MESH_MODEL_MESSAGE_EVENT_STRING, callResponse)
    }

    /**
     * Add a Vendor plugin call to the list of plugin calls to watch for a response.
     *
     * @param modelId Model ID.
     * @param opCode Operation code sent.
     * @param opPairCode Operation code to watch for.
     * @param meshAddress Mesh address.
     * @param call Plugin call.
     */
    fun addVendorPluginCall(modelId: Int, opCode: Int, opPairCode: Int, meshAddress: Int, call: PluginCall) {
        if (!MeshAddress.isValidUnicastAddress(meshAddress)) {
            return call.resolve()
        }
        pluginCalls.add(VendorPluginCall(modelId, opCode, opPairCode, meshAddress, call))
    }

    /**
     * Resolve a Vendor plugin call.
     * If the call is not found in the list of plugin calls, a notification is sent to the listeners.
     *
     * @param meshMessage Mesh message.
     */
    fun resolveVendorPluginCall(meshMessage: MeshMessage) {
        Log.d(tag, "resolveVendorPluginCall ${meshMessage.opCode} from ${meshMessage.src}")
        val callResponse = generateVendorPluginCallResponse(meshMessage)

        val pluginCall =
            pluginCalls.find { it is VendorPluginCall && it.meshOperationCallback == meshMessage.opCode && (it.meshAddress == meshMessage.src || it.meshAddress == 0xFFFF) }

        Log.d(tag, "resolveVendorPluginCall: registered call: ${pluginCall != null}")

        if (pluginCall != null) {
            pluginCall as VendorPluginCall
            pluginCall.resolve(callResponse)
            pluginCalls.remove(pluginCall)
        }
        NotificationManager.getInstance().sendNotification(MESH_MODEL_MESSAGE_EVENT_STRING, callResponse)
    }
}

File: ios/Sources/BluetoothMeshPlugin/plugin/PluginCallManager.swift
//
//  PluginCallManager.swift
//  BluetoothMeshPlugin
//
//  Created by LE BRIS Loris on 30/10/2025.
//

import Capacitor
import Foundation
import NordicMesh

public class PluginCallManager {
    static let shared = PluginCallManager()

    private var pluginCalls: [BasePluginCall] = []

    public func removePluginCall(_ call: BasePluginCall) {
        if let index = pluginCalls.firstIndex(where: { $0 === call }) {
            pluginCalls.remove(at: index)
        }
    }

    public func addSigPluginCall(_ meshOperation: UInt32, _ meshAddress: UInt16, _ call: CAPPluginCall) {
        let operationPair = OperationPairs.getMeshOperationPair(meshOperation)
        pluginCalls.append(
            SigPluginCall(operationPair, meshAddress, call)
        )
    }

    public func resolveSigPluginCall(_ response: RoutedMeshMessage) {
        let callResponse = SigPluginCall.generateSigPluginCallResponse(response)

        let index = pluginCalls.firstIndex { call in
            guard let c = call as? SigPluginCall else { return false }
            return c.meshOperationCallback == response.message.opCode && c.meshAddress == response.src
        }

        if let index = index, let sigCall = pluginCalls[index] as? SigPluginCall {
            pluginCalls.remove(at: index)
            sigCall.resolve(callResponse)
        }

        NotificationManager.shared.sendNotification(
            event: "meshModelMessageEvent",
            data: callResponse
        )
    }

    public func addConfigPluginCall(
        _ meshOperation: UInt32,
        _ meshAddress: UInt16,
        _ call: CAPPluginCall,
        timeout: TimeInterval? = nil
    ) {
        let operationPair = OperationPairs.getMeshOperationPair(meshOperation)

        let pluginCall = FoundationPluginCall(
            operationPair,
            meshAddress,
            call,
            timeout: timeout
        )

        pluginCalls.append(pluginCall)
    }

    public func resolveConfigPluginCall(_ response: RoutedMeshMessage) {
        let callResponse = FoundationPluginCall.generateFoundationPluginCallResponse(response)

        let index = pluginCalls.firstIndex { call in
            guard let c = call as? FoundationPluginCall else { return false }
            return c.meshOperationCallback == response.message.opCode && c.meshAddress == response.src
        }

        if let index = index, let configCall = pluginCalls[index] as? FoundationPluginCall {
            pluginCalls.remove(at: index)
            configCall.resolve(callResponse)
        }

        NotificationManager.shared.sendNotification(
            event: "meshModelMessageEvent",
            data: callResponse
        )
    }

    public func addVendorPluginCall(
        _ modelId: UInt32,
        _ opCode: UInt32,
        _ opPairCode: UInt32,
        _ meshAddress: UInt16,
        _ call: CAPPluginCall,
        timeout: TimeInterval? = nil
    ) {
        pluginCalls.append(VendorPluginCall(modelId, opCode, opPairCode, meshAddress, call, timeout: timeout))
    }

    public func resolveVendorPluginCall(_ response: RoutedMeshMessage) {
        let callResponse = VendorPluginCall.generateVendorPluginCallResponse(response)

        let index = pluginCalls.firstIndex { call in
            guard let c = call as? VendorPluginCall else { return false }
            return c.meshOperationCallback == response.message.opCode
                && (c.meshAddress == response.src || c.meshAddress == Address.allNodes)
        }

        if let index = index, let vendorCall = pluginCalls[index] as? VendorPluginCall {
            pluginCalls.remove(at: index)
            vendorCall.resolve(callResponse)
        }

        NotificationManager.shared.sendNotification(
            event: "meshModelMessageEvent",
            data: callResponse
        )
    }
}

extension PluginCallManager: MeshNetworkDelegate {
    public func meshNetworkManager(
        _ manager: MeshNetworkManager,
        didReceiveMessage message: MeshMessage,
        sentFrom source: Address,
        to destination: MeshAddress
    ) {
        print("Received message \(message) from \(source) to \(destination)")

        let routedMessage = RoutedMeshMessage(message: message, src: source, dst: destination)

        switch message {
        case is ConfigNodeResetStatus,
            is ConfigModelAppStatus,
            is ConfigAppKeyStatus,
            is ConfigAppKeyList,
            is ConfigCompositionDataStatus,
            is HealthFaultStatus,
            is HealthCurrentStatus:
            self.resolveConfigPluginCall(routedMessage)
            break

        case is GenericOnOffStatus,
            is GenericPowerLevelStatus,
            is LightHSLStatus,
            is LightCTLStatus,
            is LightCTLTemperatureRangeStatus:
            self.resolveSigPluginCall(routedMessage)
            break

        case is VendorResponse, is UnknownMessage:
            self.resolveVendorPluginCall(routedMessage)
            break

        default:
            break
        }

        if message is ConfigNodeResetStatus {
            NodesOnlineStateManager.shared.removeNode(unicastAddress: source)
        }
    }

    public func meshNetworkManager(
        _ manager: MeshNetworkManager,
        failedToSendMessage message: MeshMessage,
        from localElement: Element,
        to destination: MeshAddress,
        error: Error
    ) {
        print("Failed to send message \(message) from \(localElement) to \(destination): \(error)")
    }
}

public struct RoutedMeshMessage {
    public let message: MeshMessage
    public let src: Int
    public let dst: Int

    public init(message: MeshMessage, src: Address, dst: MeshAddress) {
        self.message = message
        self.src = Int(src)
        self.dst = Int(dst.address)
    }
}
